src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -163,63 +169,28 @@ struct TensorContract
 /*!
  * \ingroup TensorExpressionsGroup
  */
-template <int Index1, int Index2, typename T, typename X, typename Symm,
-          typename IndexList, typename Args>
+template <size_t ReplacedArg1, size_t ReplacedArg2, typename T, typename X,
+          typename Symm, typename IndexList, typename Args>
 SPECTRE_ALWAYS_INLINE auto contract(
     const TensorExpression<T, X, Symm, IndexList, Args>& t) {
-  return TensorContract<tmpl::int32_t<Index1>, tmpl::int32_t<Index2>, T, X,
-                        Symm, IndexList, Args>(~t);
+  return TensorContract<tmpl::size_t<ReplacedArg1>,
+                        tmpl::size_t<ReplacedArg2>, T, X, Symm, IndexList,
+                        Args>(~t);
 }
 
 namespace detail {
 // Helper struct to allow contractions by using repeated indices in operator()
 // calls to tensor.
-template <template <typename> class TE, typename ReplacedArgList, typename I,
-          typename TotalContracted>
-struct fully_contract_helper {
-  template <typename T>
-  SPECTRE_ALWAYS_INLINE static constexpr auto apply(const T& t) -> decltype(
-      contract<
-          tmpl::index_of<ReplacedArgList, ti_contracted_t<I::value>>::value,
-          tmpl::index_of<ReplacedArgList,
-                         ti_contracted_t<I::value + 1>>::value>(
-          fully_contract_helper<TE, ReplacedArgList, tmpl::next<I>,
-                                TotalContracted>::apply(t))) {
-    return contract<
-        tmpl::index_of<ReplacedArgList, ti_contracted_t<I::value>>::value,
-        tmpl::index_of<ReplacedArgList, ti_contracted_t<I::value + 1>>::value>(
-        fully_contract_helper<TE, ReplacedArgList, tmpl::next<I>,
-                              TotalContracted>::apply(t));
-  }
-};
-
-template <template <typename> class TE, typename ReplacedArgList,
-          typename TotalContracted>
-struct fully_contract_helper<TE, ReplacedArgList,
-                             tmpl::int32_t<TotalContracted::value - 1>,
-                             TotalContracted> {
-  using I = tmpl::int32_t<TotalContracted::value - 1>;
-  template <typename T>
-  SPECTRE_ALWAYS_INLINE static constexpr auto apply(const T& t) -> decltype(
-      contract<
-          tmpl::index_of<ReplacedArgList, ti_contracted_t<I::value>>::value,
-          tmpl::index_of<ReplacedArgList,
-                         ti_contracted_t<I::value + 1>>::value>(
-          TE<ReplacedArgList>(t))) {
-    return contract<
-        tmpl::index_of<ReplacedArgList, ti_contracted_t<I::value>>::value,
-        tmpl::index_of<ReplacedArgList, ti_contracted_t<I::value + 1>>::value>(
+template <template <typename> class TE, typename ReplacedArgList, size_t I,
+          typename TotalContracted, typename T>
+SPECTRE_ALWAYS_INLINE static constexpr auto fully_contract(const T& t) {
```

Comment:
I'm struggling to piece together what this is doing. For my understanding, could you describe it in words here as a Github comment? E.g., what are the template parameters and what is the `if .. else` checking for? What are typical inputs and outputs?

Reply:
Yes!

**Template paramaters**
`TE`: type of the `TensorExpression` that contains the `Tensor` to be contracted (specifically, [this](https://github.com/sxs-collaboration/spectre/blob/b9591b16b7445300026882c6e86edb5167d0262b/src/DataStructures/Tensor/Tensor.hpp#L127))

`ReplacedArgsList`: a list of `TensorIndex`s where the indices that are to be contracted have been replaced with values that follow a specific pattern. This marking is done to more easily keep track of indices to contract, and also to distinguish the upper and lower indices from each other, as they have the same `TensorIndex<#>` type.

*Example:*
If we do `R(ti_I, ti_j, ti_i)`, the corresponding list of `TensorIndex` types is `tmpl::list<TensorIndex<8>, TensorIndex<9>, TensorIndex<8>>`. `ReplacedArgList` is this list, but with the contracted index values replaced to be 1000 and 1001. So, `ReplacedArgList` in this example would be `tmpl::list<TensorIndex<1000>, TensorIndex<9>, TensorIndex<1001>>`.

If we have something like `R(ti_I, ti_j, ti_J, ti_i)`, we have 2 pairs of indices to contract, and values for the second pair will be 1002 and 1003 (carrying on after the 1000 and 1001). The corresponding list of `TensorIndex` types is `tmpl::list<TensorIndex<8>, TensorIndex<9>, TensorIndex<9>, TensorIndex<8>>` and so `ReplacedArgList` in this example would be `tmpl::list<TensorIndex<1000>, TensorIndex<1002>, TensorIndex<1003>, TensorIndex<1001>>`. Although it sounds like SpECTRE doesn't support rank 5 tensors, this should generalize if it ever needed to generalize to more than 2 contractions.

`I`: this is simply a counter that increases by 2 each time `fully_contract` is called. It's used to recreate these contracted index values of 1000 and 1001 for the first pair, and then 1002 and 1003 for the next pair. It's increased by 2 for successive calls to `fully_contract` because after one contraction, you've contracted the 1000 and 1001 indices (`ti_contracted_t<0>` and `ti_contracted_t<1>`), and if you happen to have one more, then you want to contract the next two contracted indices, 1002 and 1003 (`ti_contracted_t<2>` and `ti_contracted_t<3>`).

`TotalContracted`: the total number of pairs of indices to contract. e.g. `TotalContracted` for `R(ti_a, ti_A, ti_b)` would be 1.

`T`: the type of the `Tensor` we're wanting to contract

**On the `if`/`else`:**
The short answer is that the `if` and `else` enable recursively contracting when we need to contract more than one index pair, where the `if` is the base case. The `else` is only ever run if there is more than one pair of indices to contract, and the `if` is run when there is only one more pair left to contract. So if I had something like `R(ti_A, ti_a, ti_b, ti_c)`, then there is one pair of indices to contract. So when `fully_contract` is called [here](https://github.com/sxs-collaboration/spectre/blob/b9591b16b7445300026882c6e86edb5167d0262b/src/DataStructures/Tensor/Tensor.hpp#L282), only the `if` will be run, which simply returns a `TensorContract` object by passing the tensor to `contract`. However, if I had something like `R(ti_A, ti_a, ti_B, ti_b)`, there are two pairs to contract. When `fully_contract` is called (from the same line), the `else` block will first be entered, which also calls `contract`, but it passes to it another call to `fully_contract` that will enter the base case (the `if`). This effectively means that we create and return nested `TensorContract`s - one outer one that contains another `TensorContract`, and each of these two handle contracting one of the pairs to be contracted. It's like contracting each pair at a time, recursively - as if you first contract R_AaBb to R_Aa, and then contract the result to the scalar, R. Let me know if you're looking for more under the hood details on this.

Reply:
This was super helpful, thanks again for a clear explanation!

I think it would be valuable to document this knowledge in the code somewhere, perhaps as easily as pasting this comment in to give an overview of the whole contraction machinery. Maybe there's documentation of this somewhere already, but if so I couldn't find it...

I don't think we should hold up this PR, but something to consider doing in a future one!

Reply:
Will do in a future PR!